util/sync/mutex/keyedmutex.go

```diff
@@ -0,0 +1,77 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package mutex implements utilities for mutexes.
+package mutex
+
+import "sync"
+
+// KeyedMutex is a mutex locking on the key provided to the Lock function.
+// Only one caller can hold the lock for a specific key at a time.
+type KeyedMutex struct {
+	locksMtx sync.Mutex
+	locks    map[interface{}]*keyLock
```

Comment:
Q: should we make this more typed and just use a string as the key? Usually that is enough to fulfill a lot of use cases.